Support changing the template engine, includes an example

Jackbennett opened the pull request:
I've put the engine change as a separate commit to the inclusion of the template in case you want to cherry-pick only that change.

The example documentation on the site depends on my PR #4  being included. 

I don't know if this is the best way to do it but this is one way I got it working which I wanted to try before just creating an issue.

I've done some testing I can think of to make it work but this could probably use some actual written tests.

franciscop commented:
Hi @Jackbennett , thanks for the PRs!

Right now the library is still quite experimental, so there are things where I don't want to commit yet and other things that are contradicting each other in what I wrote so far.

About this PR, the options.md file that you set should render here:

> [https://serverjs.io/documentation/server/](https://serverjs.io/documentation/server/)

So you would have to edit it in here:

> [https://github.com/franciscop/server/blob/master/docs/documentation/server/README.md](https://github.com/franciscop/server/blob/master/docs/documentation/server/README.md)

As you noted, the root folder `/documentation` should be removed and instead use `/docs/documentation` since the folder `/docs` is used for generating the web.

I keep what I said about committing too soon; I don't really like this:

```js
server({
  ...
  'view engine': 'handlebars',
  engine: { handlebars: exphbs() }
}, get(...));
```

So just setting this would be preferable:

```js
server({
  ...
  engine: { handlebars: exphbs() }
}, get(...));
```

Which is totally doable (just add an extra `app.set()` within that `if` in your PR).

> Note that you might not even need to specify the function: http://expressjs.com/en/guide/using-template-engines.html

While this would have to be tested properly, I can see how we could instead have a dual situation:

```js
const options = { engine: 'pug' };
const options = { engine: { pug: whatever() } };
```

In this way we don't have to confuse the user with `view engine` vs `engine` vs `npm install ENGINE --save` vs [...] and we can focus on a single option parameter and explanation.

Conclusion:

- The Options file is good but should be in a different place. I will try to clean up the folder/file mess soon in case you want to help somewhere else.
- I like the engine idea, but it would need some extra work [to align with *server*'s objectives and goals](https://serverjs.io/about).

franciscop commented:
Didn't mean to close it, sorry

Jackbennett commented:
Thanks for the feedback.

```js
const options = { engine: 'pug' };
const options = { engine: { pug: whatever() } };
```

So I'll probably make a new Pr for the corresponding answer but to align with the goals; to do the above which I agree looks good would you be removing the 'view engine' property as it's redundant? Though it's expresses property so that's probably not a good idea.

Options seem to me;
1. `view engine` || `engine` accept string or object, choosing one to take precedence and document that you don't set both.
2. `view engine` accepts string or object. this hides `app.engine` from a user. which isn't bad.
3. Invert option 2. hiding `view engine` for `engine`

I've done some testing and option 2 looks the cleanest. I will either update this PR or make a new one depending how how well I know how to use git lol.

re: docs, I missed that options.md was already appended into server/readme.md so I thought it was a sub-page that wasn't finished/draft not yet included. nevermind.

franciscop commented:
Though it uses express in the back and I will probably never remove that dependency, the main objective is that the devs don't need to know how to use `express` to use `server` (in fact that would be kind of against the objectives of `server`).

So the idea is:

- 50+%* of the devs don't even need to set-up `engine`
- 95+% of the devs can work with the default or just do it with a string like `{ engine: 'pug' }`
- 99+% of the devs can work with the default, the string or the object like `{ engine: { pug: whatever() } }`

*I say 50% as the engine is something quite personal, this number for other options like the port should be ~90%

The 0.x% of the devs who really need more flexibility or something more powerful than that can *probably* dig into the code, ask questions in SO or handle it. These devs probably aren't *beginners* nor *people launching many projects with the defaults*, for which *express*, *yeoman* or another tool might be better suited.

So just doing `{ engine: 'pug' }` should work for most people; I am open to both `engine` vs `view engine` (but I like `engine` a bit more since it's not the same as `view engine` in express).

franciscop commented:
About git, feel free to close/open another, I also get confused all the time when I have to do something complex.

Re: I just appended it! Sorry for not waiting but it felt trivial (:

Jackbennett commented:
Right I've been toiling with git so I'd not read any response by the time I've already updated this.

The problem I had with `engine` is I would get this.engine is not a function as I think app.set('engine',... was getting evaluated by express anyway. This was some hours ago now I've replaced that thought with how git rebase works sorry. 

If you want this seems quite neat and tidy. I remove the docs update from this PR so if you're happy with this I'll write something. 

Could use some tests however I'm sure.

I couldn't get this to work with just `'view engine': 'handlebars'` it doesn't seem to find the right function from express-handlebars but that could just be my inexperience.

franciscop commented:
Maybe this helps: https://webapplog.com/jade-handlebars-express/

Jackbennett commented:
yeh it does haha 
>The express3-handlebars approach usage is as follows:
```
app.engine('handlebars', exphbs({defaultLayout: 'main'}))
```

so yeh the original example that's the way to do it. The [hbs](https://github.com/pillarjs/hbs) wrapper has more recent changes maybe, using that instead express4 find it so you don't even need the engine change. But I guess nice to know it's there after all.

franciscop commented:
I think I messed things up with git and cannot handle it right now, will try to fix it tonight or tomorrow. Sorry!

Jackbennett commented:
Looks like you've completely replaced the repo. Don't forget about my fork if you end up needing the old history back. The #github freenode irc were very helpful to me changing this PR
